### ISSUE TITLE
Additional interface info?

### ISSUE THREAD

**Issue by QuantumEntangledAndy** [EstebanBorai/local-ip-address#26]
**Feature request**

Do you think it would be possible add getting additional interface information such as subnet mask and the broadcast address?

**Comment by EstebanBorai** [comment]
Hi @QuantumEntangledAndy!

Thanks for the interest in this crate, yes I'am designing an API at the moment, there's very good insights so far including this feature request from you.

As of now I have this structure which I discussed already with @zyddnys here: https://github.com/EstebanBorai/local-ip-address/issues/24

```rust
enum NetworkInterfaceAddress {
    AfInet,
    AfInet6,
}

struct NetworkInterface {
    name: String,
    ip_address: Option<IpAddr>,
    family: NetworkInterfaceFamily,
}
```

For this feature request I would have to define these types (subnet mask and broadcast address) to make them compatible with macOS, Linux and Windows (in order to offer the same API).

Any ideas/contributions are welcome!

Do you have an idea of how we could extend the current draft to support these fields you are requesting?

**Comment by QuantumEntangledAndy** [comment]
Perhaps follow the scheme in this [crate](https://docs.rs/get_if_addrs/0.5.3/get_if_addrs/)

So netmask is an `ipv4/6` and broadcast is an `Option<ipv4/6>`

**Comment by QuantumEntangledAndy** [comment]
Also how would your crate handle multiple ip addresses per interface?

**Comment by QuantumEntangledAndy** [comment]
What about something like this?

struct NetworkInterface {
    name: String,
    ip_addresses: Vec<IfAddr>,
    family: NetworkInterfaceFamily,
}

struct IfAddr {
    ip_address: IpAddr,
    subnet: IpAddr,
    broadcast: Option<IpAddr>
}

Also how do you handle interfaces with both v4 and v6 family?

**Comment by EstebanBorai** [comment]
@QuantumEntangledAndy looks good so far.

> Also how would your crate handle multiple ip addresses per interface?

We would have to provide a `Vec<IpAddr>` then, `IpAddr` already covers `IpV4` and `IpV6` so it's solved unless I'm missing something?

> Perhaps follow the scheme in this crate

Will take a look! Thanks for referencing!

**Comment by QuantumEntangledAndy** [comment]
Well ipaddr subnet and broadcast will all be ipv4 or ipv6 as a set. So could have IfAddr as an enum of ifaddr4/ifaddr6 at that level rather than at the IP level. Maybe could skip the family that way as it's encoded on the enum.

**Comment by QuantumEntangledAndy** [comment]
```rust
Enum IfAddr {
    V4(IfAddrV4),
    V6(IfAddrV6),
}

struct IfAddrV4 {
    ip_address: IpAddrV4,
    subnet: IpAddrV4,
    broadcast: Option<IpAddrV4>
}

struct IfAddrV6 {
    ip_address: IpAddrV6,
    subnet: IpAddrV6,
    broadcast: Option<IpAddrV6>
}
```

**Comment by EstebanBorai** [comment]
Hi @QuantumEntangledAndy! I have been working on the support for extra interface info here:
https://github.com/EstebanBorai/local-ip-address/tree/feat/additional-interface-info

Please let me know any suggestions and insights.
I would like to open a discussion for this so we keep all the relevant comments there, what do you think?

**Comment by EstebanBorai** [comment]
@QuantumEntangledAndy given that we are going further into different getifaddrs implementations for (as of today) 3 target operative systems, I will be moving the logic to handle network interfaces to its own crate and will have this crate for local-ip-address only (as the name suggest), which will consume this other getifaddrs-based crate to gather host's local IP address.

**Comment by QuantumEntangledAndy** [comment]
I see I'll look out for it then thanks for the heads up.

**Comment by EstebanBorai** [comment]
> I see I'll look out for it then thanks for the heads up.

Okey! Here it is: https://github.com/EstebanBorai/network-interface
So this issue basically turned into a crate (?).

I've moved my work on the branch I been working on here wrt additional interface data into that crate so feel free to open issues or we could have a discussion with RFC for version 0.1.0.

I've published a v0.1.0-beta so we can install it and play around easily.

Thanks again @QuantumEntangledAndy! Hope to see you there!
